### PR TITLE
fix: add types for `ElementRef` support

### LINKED ISSUE #225: Can't get `ElementRef` type from `styled` comp
The following `CardElement` type will be `never` but ideally would be `ReactNative.View`.

```jsx
const Card = styled(ReactNative.View)({});

// Later 
type CardElement = React.ElementRef<typeof Card>;
```

Here's an example that works as expected without the `styled` wrapper:

image

### PR DESCRIPTION
**Pull request by jjenzz**
> **Note**: Easier to review with hidden whitespace changes 

Closes #225

### CODE REVIEW

File: packages/core/src/css/styled.tsx [hunk 1]
```diff
@@ -17,14 +17,10 @@ import type { ThemedOptions } from './types'
  *
  */
 export function styled<
-  Props,
-  ThemeKey extends keyof DripsyFinalTheme = keyof DripsyFinalTheme,
-  P extends Omit<Props, 'variant' | 'variants'> = Omit<
-    Props,
-    'variant' | 'variants'
-  >
```
Comment (jjenzz) [inline packages/core/src/css/styled.tsx:25]: moved here https://github.com/nandorojo/dripsy/pull/237/files#diff-8ce561232f7fced53e331bcb44b040c8051bc86a90780b86793a261a6b328aaaR10-R13

**Comment by nandorojo** [comment]
This is 🔥🔥🔥

I never knew how to get inferred refs to get to work. I'd love to get this to work for `moti` [too](https://github.com/nandorojo/moti/blob/master/packages/core/src/motify.tsx#L33-L39). I'll try to copy this over there.

**Comment by nandorojo** [comment]
One thing – I'm getting this error when I run `yarn` (which triggers `yarn prepare`):

```ts
src/css/create-themed-component.tsx(48,52): error TS2345: Argument of type 'PropsWithChildren<StyledProps<ThemeKey> & Props>' is not assignable to parameter of type 'ExtraProps'.
  'ExtraProps' could be instantiated with an arbitrary type which could be unrelated to 'PropsWithChildren<StyledProps<ThemeKey> & Props>'.
```

**Comment by nandorojo** [comment]
I added a commit to ignore it – I think it's fine since it's just an implementation detail.

**Comment by jjenzz** [comment]
i'll take a quick look at that issue, i think i know the cause...

**Comment by jjenzz** [comment]
@nandorojo all should be good to go now :)

**Comment by jjenzz** [comment]
hmmm, i've pulled the latest version and for some reason the following code is fine in this project but errors in mine. does this work for you?

```jsx
interface CircleProps {
  size?: number
  color?: 'red' | 'green'
}

const Circle = styled(View)<CircleProps>((props) => ({
  width: props.size || 56,
  height: props.size || 56,
  backgroundColor: props.color ?? 'transparent',
  borderRadius: '$rounded',
  justifyContent: 'center',
  alignItems: 'center',
  overflow: 'hidden',
}))

export const Foo = () => {
  return (
    <Circle
      size={20}
      sx={{
        backgroundColor: '$primary',
        borderWidth: 2,
        borderColor: '$background',
        position: 'absolute',
        right: -6,
        bottom: 2,
      }}
    />
  )
}
```

**Comment by nandorojo** [comment]
Oh, it's because the first generic of `styled` became `C`, when it used to be `Props`, right?

**Comment by nandorojo** [comment]
```ts
const Circle = styled(View)((props: CircleProps) => ({
```

This fixes it, but it still keeps a breaking change.

**Comment by nandorojo** [comment]
I think this should fix it:

```tsx
export function styled<
  Props,
  C extends ComponentType<any>,
  ThemeKey extends keyof DripsyFinalTheme = keyof DripsyFinalTheme
>(
  Component: C,
  {
    themeKey,
    defaultVariant,
  }: Pick<ThemedOptions<any, ThemeKey>, 'themeKey' | 'defaultVariant'> = {}
) {
  return function dripsyFactory<Extra>(
    defaultStyle?: ThemedOptions<Extra & Props, ThemeKey>['defaultStyle']
  ) {
    return createThemedComponent<C, Extra & Props, ThemeKey>(Component, {
      defaultVariant,
      themeKey,
      defaultStyle,
    })
  }
}
```

**Comment by nandorojo** [comment]
Hm, that actually did not fix it for me. I'll have to keep looking.

**Pull request merged**